File: src/main/java/com/freya02/botcommands/api/core/service/annotations/Resolver.java
package com.freya02.botcommands.api.core.service.annotations;

import com.freya02.botcommands.api.parameters.ParameterResolver;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Marks this class as a {@link ParameterResolver parameter resolver}.
 * <br>This is a specialization of {@link BService} for parameter resolvers.
 *
 * <p><b>Requirement:</b> this must extend {@link ParameterResolver}.
 *
 * @see BService
 *
 * @see ParameterResolver
 */
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.CLASS) //Read by ClassGraph
public @interface Resolver { }

File: src/main/kotlin/com/freya02/botcommands/api/core/service/annotations/Resolver.kt
package com.freya02.botcommands.api.core.service.annotations

import com.freya02.botcommands.api.parameters.ParameterResolver

/**
 * Marks this class as a [parameter resolver][ParameterResolver].<br>
 * This is a specialization of [BService] for parameter resolvers.
 *
 * **Requirement:** this must extend [ParameterResolver].
 *
 * **Warning:** Top-level functions are not processed, you must have them in an object/class.
 *
 * @see BService @BService
 *
 * @see ParameterResolver
 */
@Target(AnnotationTarget.CLASS, AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER)
@Retention(AnnotationRetention.BINARY) //Read by ClassGraph
annotation class Resolver